noether_filtering/cloud_filter_plugins.xml

```diff
@@ -0,0 +1,129 @@
+    </description>
+  </class>
+  <class name="noether_filtering::cloud::CropBoxFilter&lt;pcl::PointNormal>" type="noether_filtering::cloud::CropBoxFilter&lt;pcl::PointNormal>"
+    base_class_type="noether_filtering::FilterBase&lt;boost::shared_ptr&lt;pcl::PointCloud&lt;pcl::PointNormal> > >">
+    <description>
+      Crop box filter for PointNormal type
+    </description>
+  </class>
+
+  <!-- pcl::PointXYZRGB -->
+  <class name="noether_filtering::cloud::VoxelGridFilter&lt;pcl::PointXYZRGB>" type="noether_filtering::cloud::VoxelGridFilter&lt;pcl::PointXYZRGB>"
+    base_class_type="noether_filtering::FilterBase&lt;boost::shared_ptr&lt;pcl::PointCloud&lt;pcl::PointXYZRGB> > >">
+    <description>
+      Voxel grid filter for PointXYZRGB type
+    </description>
+  </class>
+  <class name="noether_filtering::cloud::StatisticalOutlierFilter&lt;pcl::PointXYZRGB>" type="noether_filtering::cloud::StatisticalOutlierFilter&lt;pcl::PointXYZRGB>"
+    base_class_type="noether_filtering::FilterBase&lt;boost::shared_ptr&lt;pcl::PointCloud&lt;pcl::PointXYZRGB> > >">
+    <description>
+      Statistical outlier filter for PointXYZRGB type
+    </description>
+  </class>
+  <class name="noether_filtering::cloud::PassThroughFilter&lt;pcl::PointXYZRGB>" type="noether_filtering::cloud::PassThroughFilter&lt;pcl::PointXYZRGB>"
+    base_class_type="noether_filtering::FilterBase&lt;boost::shared_ptr&lt;pcl::PointCloud&lt;pcl::PointXYZRGB> > >">
+    <description>
+      Pass through filter for PointXYZRGB type
+    </description>
+  </class>
+  <class name="noether_filtering::cloud::RadiusOutlierFilter&lt;pcl::PointXYZRGB>" type="noether_filtering::cloud::RadiusOutlierFilter&lt;pcl::PointXYZRGB>"
+    base_class_type="noether_filtering::FilterBase&lt;boost::shared_ptr&lt;pcl::PointCloud&lt;pcl::PointXYZRGB> > >">
+    <description>
+      Radius outlier filter for PointXYZRGB type
+    </description>
+  </class>
+  <class name="noether_filtering::cloud::CropBoxFilter&lt;pcl::PointXYZRGB>" type="noether_filtering::cloud::CropBoxFilter&lt;pcl::PointXYZRGB>"
+    base_class_type="noether_filtering::FilterBase&lt;boost::shared_ptr&lt;pcl::PointCloud&lt;pcl::PointXYZRGB> > >">
+    <description>
+      Crop box filter for PointXYZRGB type
+    </description>
+  </class>
+
+  <!-- pcl::PointXYZI -->
+  <class name="noether_filtering::cloud::VoxelGridFilter&lt;pcl::PointXYZI>" type="noether_filtering::cloud::VoxelGridFilter&lt;pcl::PointXYZI>"
+    base_class_type="noether_filtering::FilterBase&lt;boost::shared_ptr&lt;pcl::PointCloud&lt;pcl::PointXYZI> > >">
+    <description>
+      Voxel grid filter for PointXYZI type
+    </description>
+  </class>
+  <class name="noether_filtering::cloud::StatisticalOutlierFilter&lt;pcl::PointXYZI>" type="noether_filtering::cloud::StatisticalOutlierFilter&lt;pcl::PointXYZI>"
+    base_class_type="noether_filtering::FilterBase&lt;boost::shared_ptr&lt;pcl::PointCloud&lt;pcl::PointXYZI> > >">
```

Comment:
I just realized the the output arguments to the plugin is a shared_ptr reference, in general it isn't good practice to pass a pointer by reference, there are some articles, threads that discuss this at length [here](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-summary-smartptrs) and [here](https://stackoverflow.com/questions/8385457/should-i-pass-a-shared-ptr-by-reference)
Could we pass references instead, at least of the non-const output argument?

Reply:
I'm passing pointer types here because all of the PCL filters I've run into take shared pointers as an argument to their `setInputCloud` method. If we passed references of clouds instead, I would have to allocate a copied input cloud on the heap and make a shared pointer to it to be able to use the filter (unless there is a way to make a shared pointer to a stack-allocated object, which I don't think is possible)

Reply:
Perhaps we could modify the fiter base class to accept an additional output template argument, so it would be
```
template <typename T, template W>
class FilterBase
```
for cloud filters you would then create a filter class with the following template values and a `filter` override method
```
template <typename T>
class CloudFilter: public FilterBase<Cloud::Ptr, Cloud>
{
  .
  .
  . 
 virtual bool filter(const Cloud::Ptr& input, Cloud& output) final // overrides default filter method
```

Reply:
We could do that, but I think I found a way to get around this issue. You can actually create a shared pointer from a reference to a stack-allocated object and give it a "destructor" function that doesn't actually delete the object. I'll add this change into the cloud filters and change their filter signature accordingly